src/cls/rgw/cls_rgw.cc

```diff
@@ -1224,11 +1352,36 @@ int rgw_bucket_complete_op(cls_method_context_t hctx, bufferlist *in, bufferlist
 } // rgw_bucket_complete_op
 
 template <class T>
-static int write_entry(cls_method_context_t hctx, T& entry, const string& key)
+static int write_entry(cls_method_context_t hctx, T& entry, const string& key,
+                       rgw_bucket_dir_header& header)
 {
   bufferlist bl;
   encode(entry, bl);
-  return cls_cxx_map_set_val(hctx, key, &bl);
+  int ret = cls_cxx_map_set_val(hctx, key, &bl);
+  if (ret < 0) {
+    return ret;
+  }
+  if (header.resharding_in_logrecord()) {
+    ret = reshard_log_index_operation(hctx, key, entry.key, &bl);
+    header.reshardlog_entries++;
+  }
+  return ret;
+}
+
+static int remove_entry(cls_method_context_t hctx, const string& idx,
+                        const cls_rgw_obj_key& key,
+                        rgw_bucket_dir_header& header)
+{
+  int ret = cls_cxx_map_remove_key(hctx, idx);
+  if (ret < 0) {
+    CLS_LOG(0, "ERROR: cls_cxx_map_remove_key() idx=%s ret=%d", idx.c_str(), ret);
+    return ret;
+  }
+  if (header.resharding_in_logrecord()) {
+    bufferlist empty;
+    return reshard_log_index_operation(hctx, idx, key, &empty);
```

Comment:
should we increment `header.reshardlog_entries` here, like to we do in `write_entry()`?

Reply:
@cbodley  Because reshardlog_entries counts in one reshard process, and it will reset to 0 when canceling a reshard too. Though log entries will be increased when removing entry, original index entries will be decreasing meantime. So I think there is no need to add it in this place. Is it reasonable?
If you think increment is necessary, I will do it.

Reply:
done